web/packages/teleport/src/ThemeProvider.tsx

```diff
@@ -0,0 +1,125 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React, { useState, useEffect, ReactNode } from 'react';
+import { StaticThemeProvider } from 'design/ThemeProvider';
+
+import { Theme as ThemePreference } from 'gen-proto-ts/teleport/userpreferences/v1/theme_pb';
+
+import { bblpTheme, lightTheme, darkTheme, Theme } from 'design/theme';
+
+import cfg from 'teleport/config';
+import { storageService, KeysEnum } from 'teleport/services/storageService';
+
+export const ThemeProvider = (props: { children?: ReactNode }) => {
+  const [themePreference, setThemePreference] = useState<ThemePreference>(
+    storageService.getThemePreference()
+  );
+
+  useEffect(() => {
+    storageService.subscribe(receiveMessage);
+
+    function receiveMessage(event) {
+      const { key, newValue } = event;
+
+      if (!newValue || key !== KeysEnum.USER_PREFERENCES) {
+        return;
+      }
+
+      const preferences = JSON.parse(newValue);
+      if (
+        preferences.theme !== ThemePreference.UNSPECIFIED &&
+        preferences.theme !== themePreference
+      ) {
+        setThemePreference(preferences.theme);
+      }
+    }
+
+    // Cleanup on unmount
+    return function unsubscribe() {
+      storageService.unsubscribe(receiveMessage);
+    };
+  }, [themePreference]);
+
+  const customThemes = {
+    bblp: bblpTheme,
+  };
+
+  // If no props.theme is defined, use the custom theme instead of the user preference theme.
+  let theme: Theme;
+  if (customThemes[cfg.customTheme]) {
+    theme = customThemes[cfg.customTheme];
+  } else {
+    theme = themePreferenceToTheme(themePreference);
+  }
+
+  return (
+    <StaticThemeProvider theme={theme}>{props.children}</StaticThemeProvider>
+  );
+};
+
+function themePreferenceToTheme(themePreference: ThemePreference): Theme {
+  if (themePreference === ThemePreference.UNSPECIFIED) {
+    return getPrefersDark() ? lightTheme : darkTheme;
+  }
+  return themePreference === ThemePreference.LIGHT ? lightTheme : darkTheme;
+}
+
+// because unspecified can exist but only used as a fallback and not an option,
+// we need to get the current/next themes with getPrefersDark in mind.
+// TODO (avatus) when we add user settings page, we can add a Theme.SYSTEM option
+// and remove the checks for unspecified
+export function getCurrentTheme(
+  currentTheme: ThemePreference
+): ThemePreference {
+  if (currentTheme === ThemePreference.UNSPECIFIED) {
+    return getPrefersDark() ? ThemePreference.DARK : ThemePreference.LIGHT;
+  }
+
+  return currentTheme;
+}
+
+export function getNextTheme(currentTheme: ThemePreference): ThemePreference {
```

Comment:
```suggestion
export function switchTheme(currentTheme: ThemePreference): ThemePreference {
```

or maybe even `switchThemeMode` ? Otherwise I could not understand what `Next` meant without reading the return below.

(also JSDoc here and below in this file.)

Reply:
This function is just kind of bad and a bandaid until the user preferences page goes live. It doesn't actually switch theme but just tells us what the "next" theme would be. It's used to determine what text will be shown in the "switch to light/dark theme" menu.

Reply:
I kept `getNextTheme` because of this ⬆️